### PR TITLE
fix(StreamID): wrap quinn::StreamID to avoid misleading log

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/connection.rs [hunk 1]
```diff
@@ -186,6 +186,88 @@ impl fmt::Debug for Connection {
...
+            dir,
+            self.index()
+        )
+    }
+}
+
+impl StreamId {
+    /// Create a new StreamId
+    pub fn new(initiator: Side, dir: Dir, index: u64) -> Self {
+        StreamId(index << 2 | (dir as u64) << 1 | initiator as u64)
```
Comment (bochaco) [inline src/connection.rs:247]: Why don't we simply use the enums instead of doing this type of encoding of these values?

**Review by bochaco — COMMENTED** [review]
I'm also thinking we could use the conn id to also be part of the stream id. E.g. I'm seeing some connection to a client is closed and later on I see we fail to send a msg to that client over a stream, but I have not way to tell if the stream that failed was over the connection that I saw being closed beforehand.

File: src/connection.rs [hunk 2]
```diff
@@ -186,6 +186,88 @@ impl fmt::Debug for Connection {
...
+    /// The initiator of a connection
+    Initiator = 0,
+    /// The acceptor of a connection
+    Acceptor = 1,
+}
+
+/// Whether a stream communicates data in both directions or only from the initiator
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum Dir {
```
Comment (bochaco) [inline src/connection.rs:202]: There is no harm in using full names for better clarity, i.e. `Direction`

File: src/connection.rs [hunk 3]
```diff
@@ -186,6 +186,88 @@ impl fmt::Debug for Connection {
     }
 }
 
+/// Whether an endpoint was the initiator of a connection
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum Side {
```
Comment (bochaco) [inline src/connection.rs:192]: Perhaps this shall be called `StreamInitiator` for better clarity?

File: src/connection.rs [hunk 4]
```diff
@@ -186,6 +186,88 @@ impl fmt::Debug for Connection {
...
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct StreamId(#[doc(hidden)] pub u64);
+
+impl fmt::Display for StreamId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let initiator = match self.initiator() {
+            Side::Initiator => "initiator",
+            Side::Acceptor => "acceptor",
+        };
+        let dir = match self.dir() {
```
Comment (bochaco) [inline src/connection.rs:230]: You don't need this, or even the `Display` impl for `Dir`, you can just use the `Debug` version of it in L236, i.e. `{dir:?}`.

File: src/connection.rs (hunk 1 above)
Comment (maqi) [reply src/connection.rs:247]: That was copied from quinn's design.
I think it's trying to encode the info into the u64

**Comment by maqi** [comment]
> I'm also thinking we could use the conn id to also be part of the stream id. E.g. I'm seeing some connection to a client is closed and later on I see we fail to send a msg to that client over a stream, but I have not way to tell if the stream that failed was over the connection that I saw being closed beforehand.

That then will require the connection_id to be passed down when contruct the stream.
will impact the API and prefer to be done as a separate PR?

**Comment by b-zee** [comment]
Is this purely for printing out a more accurate format from our perspective? Isn't it possible to simply do this?

```
struct StreamId(quinn::StreamId);
impl fmt::Display for StreamId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let initiator = match self.0.initiator() {
            quinn_proto::Side::Client => "initiator",
            quinn_proto::Side::Server => "acceptor",
        };
        let dir = match self.0.dir() {
            quinn_proto::Dir::Uni => "uni",
            quinn_proto::Dir::Bi => "bi",
        };
        write!(
            f,
            "{} {}directional stream {}",
            initiator,
            dir,
            self.0.index()
        )
    }
}
```

**Comment by maqi** [comment]
> Is this purely for printing out a more accurate format from our perspective? Isn't it possible to simply do this?
> 
> ```rust
> struct StreamId(quinn::StreamId);
> impl fmt::Display for StreamId {
>     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
>         let initiator = match self.0.initiator() {
>             quinn_proto::Side::Client => "initiator",
>             quinn_proto::Side::Server => "acceptor",
>         };
>         let dir = match self.0.dir() {
>             quinn_proto::Dir::Uni => "uni",
>             quinn_proto::Dir::Bi => "bi",
>         };
>         write!(
>             f,
>             "{} {}directional stream {}",
>             initiator,
>             dir,
>             self.0.index()
>         )
>     }
> }
> ```

yes, it is. The current logging from quinn::Streaming is bit confusing as using `Slient/Server` to describe actually the `Initiator/Acceptor` of a connection

File: src/connection.rs (hunk 4 above)
Comment (maqi) [reply src/connection.rs:230]: This is for a more concised log for the StreamId, which is different to the default Debug or Dispay impl of `enum Dir`

File: src/connection.rs (hunk 1 above)
Comment (bochaco) [reply src/connection.rs:247]: We shouldn't need such encoding here I believe?

File: src/connection.rs (hunk 4 above)
Comment (bochaco) [reply src/connection.rs:230]: I think you don't need them, let me put an example further below in the other discussion

**Comment by bochaco** [comment]
> prefer to be done as a separate PR?

I'll do that shortly

**Comment by bochaco** [comment]
> Then what about my code? Doesn't that fulfill the same purpose? 

I agree with that impl above, we shouldn't need any of the additional enums as they are just to generate that string we use for logging purposes

**Review by b-zee — APPROVED** [review]

**Pull request merged**